SI-8111 Repair symbol owners after abandoned named-/default-args

retronym opened the pull request:
Names/Defaults eagerly transforms an application with temporaries
to maintain evaluation order, and dutifully changes owners of
symbols along the way.

However, if this approach doesn't work out, we throw away this
and try a auto-tupling. However, we an still witness symbols
owned by the temporaries.

This commit records which symbols are owned by the context.owner
before `transformNamedApplication`, and rolls back the changes
before `tryTupleApply`.

Perhaps a better approach would be to separate the names/defaults
applicability checks from the evaluation-order-preserving transform,
and only call the latter after we have decided to go that way.

retronym committed:
SI-8111 Repair symbol owners after abandoned named-/default-args

Names/Defaults eagerly transforms an application with temporaries
to maintain evaluation order, and dutifully changes owners of
symbols along the way.

However, if this approach doesn't work out, we throw away this
and try a auto-tupling. However, we an still witness symbols
owned by the temporaries.

This commit records which symbols are owned by the context.owner
before `transformNamedApplication`, and rolls back the changes
before `tryTupleApply`.

Perhaps a better approach would be to separate the names/defaults
applicability checks from the evaluation-order-preserving transform,
and only call the latter after we have decided to go that way.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -3269,6 +3269,13 @@ trait Typers extends Modes with Adaptations with Tags {
+              }
+              def rollbackNamesDefaultsOwnerChanges() {
+                symsOwnedByContextOwner foreach (_.owner = context.owner)
+              }
```
Would probably be useful to have this commented

xeno-by replied:
Also, I'm having hard time understanding why this isn't a no-op. You take symbols whose owner is context.owner, and then change their owners to context.owner. What effect does this achieve?

retronym commented:
I'll add a comment. The owners are mutated in the NamesDefaults.

retronym commented:
Comment added.

Review by @xeno-by

xeno-by commented:
You alluded to the fact that it might be possible to avoid reverting. Would one of the strategies of doing that be reasonably easy to implement?

som-snytt commented:
If the principled approach that avoids undoing (there's a pithy paulpism to that effect) is not required immediately, I'll take a look.  There's another named/defaults corner case issue, so it would be fun to muck around and also redeem my regression.

retronym commented:
> You alluded to the fact that it might be possible to avoid reverting. Would one of the strategies of doing that be reasonably easy to implement?

Approach 1:

We could deeply duplicate trees, cloning local symbols, before the attempt at the names/defaults transform. We would throwaway that tree (and the orphaned symbols it contains) if we fall back to tupling. Looks like we have some infrastructure in place for this in `BodyDuplicator` used by specialization.

This might have a negative effect on performance.

Approach 2:

Untangle the logic to determine if a names/defaults application is applicable from the tree transformations. I'm having lunch with @lrytz today, so I'll ask for some insights about this approach.

Short answer: i don't think either approach is easy to implement.

xeno-by commented:
If it's too complicated for now, let's put a TODO in the code and outline possible solutions. From what I can see, this is a critical bug, so we should probably fix it by all means necessary for now, right?

retronym committed:
SI-8111 Expand the comment with a more detailed TODO

As everone knows, undo/reset/retype/rollback are bandaids; we
should try to treat the disease more directly.

retronym commented:
I marked the bug as critical on the grounds of the spectacular failure mode. But I do think the bandaid solution is a good starting point.

I've added the TODO comment.

xeno-by commented:
LGTM

xeno-by commented:
I think we could merge this one, right?
